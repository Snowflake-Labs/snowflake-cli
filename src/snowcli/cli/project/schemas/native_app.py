from __future__ import annotations

from snowcli.cli.project.schemas.relaxed_map import FilePath, Glob, RelaxedMap
from snowcli.cli.project.util import IDENTIFIER, SCHEMA_AND_NAME
from strictyaml import (
    Bool,
    Optional,
    Regex,
    Seq,
    Str,
    UniqueSeq,
)

PathMapping = RelaxedMap(
    {
        "src": Glob() | Seq(Glob()),
        Optional("dest"): FilePath(),
    }
)

native_app_schema = RelaxedMap(
    {
        "name": Str(),
        "artifacts": Seq(FilePath() | PathMapping),
        Optional("deploy_root", default="output/deploy/"): FilePath(),
        Optional("source_stage", default="app_src.stage"): Regex(SCHEMA_AND_NAME),
        Optional("package"): RelaxedMap(
            {
                Optional("scripts", default=None): UniqueSeq(FilePath()),
                Optional("role"): Regex(IDENTIFIER),
                Optional("name"): Regex(IDENTIFIER),
                Optional("warehouse"): Regex(IDENTIFIER),
            }
        ),
        Optional("application"): RelaxedMap(
            {
                Optional("role"): Regex(IDENTIFIER),
                Optional("name"): Regex(IDENTIFIER),
                Optional("warehouse"): Regex(IDENTIFIER),
                Optional("debug", default=True): Bool(),
            }
        ),
    }
)
