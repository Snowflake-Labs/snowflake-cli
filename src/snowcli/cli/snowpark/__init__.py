import typer

from snowcli.cli.snowpark.function import app as function_app
from snowcli.cli.snowpark.package import app as package_app
from snowcli.cli.snowpark.procedure import app as procedure_app

app = typer.Typer(
    name="snowpark",
    context_settings={"help_option_names": ["-h", "--help"]},
    help="Manage functions, procedures and Snowpark objects",
)

app.add_typer(function_app)
app.add_typer(package_app)
app.add_typer(procedure_app)
