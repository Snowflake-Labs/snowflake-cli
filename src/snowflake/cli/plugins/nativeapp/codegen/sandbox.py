import os
import shutil
import subprocess
import sys
import tempfile
from contextlib import contextmanager
from enum import Enum
from pathlib import Path
from typing import Optional

from click.exceptions import ClickException


class SandboxExecutionError(ClickException):
    """An error occurred while executing a python script."""

    def __init__(self, error: str):
        super().__init__(f"Failed to execute python script. {error}")


def _get_active_venv_dir() -> Optional[str]:
    return os.environ.get("VIRTUAL_ENV")


def _get_active_conda_env() -> Optional[str]:
    return os.environ.get("CONDA_DEFAULT_ENV")


def _is_venv_active() -> bool:
    return _get_active_venv_dir() is not None


def _is_conda_active() -> bool:
    return _get_active_conda_env() is not None


def _is_ms_windows() -> bool:
    return sys.platform == "win32"


@contextmanager
def _temp_script_file(script_source: str):
    script_file = tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=True)
    try:
        script_file.write(script_source)
        script_file.flush()

        yield script_file.name
    finally:
        script_file.close()


def _execute_in_venv(
    script_source: str, venv_path: Optional[Path] = None
) -> subprocess.CompletedProcess:
    resolved_venv_path = None
    if venv_path is None:
        active_venv_dir = _get_active_venv_dir()
        if active_venv_dir is not None:
            resolved_venv_path = Path(active_venv_dir).resolve()
    else:
        resolved_venv_path = venv_path.resolve()

    if resolved_venv_path is None:
        raise SandboxExecutionError("No venv root found")

    if not resolved_venv_path.is_dir():
        raise SandboxExecutionError(
            f"venv path should be an existing directory: {resolved_venv_path}"
        )

    # find the python interpreter for this environment. There is no need to activate environment prior to invoking the
    # interpreter, as venv maintains the invariant that invoking any of the scripts will set up the virtual environment
    # correctly. activation scripts are only used for convenience in interactive shells.
    if _is_ms_windows():
        python_executable = resolved_venv_path / "Scripts" / "python.exe"
    else:
        python_executable = resolved_venv_path / "bin" / "python3"

    if not python_executable.is_file():
        raise SandboxExecutionError(
            f"No venv python executable found: {resolved_venv_path}"
        )

    with _temp_script_file(script_source) as script_file:
        return subprocess.run(
            [python_executable, script_file], capture_output=True, text=True
        )


def _execute_in_conda_env(
    script_source: str, env_name: Optional[str] = None
) -> subprocess.CompletedProcess:
    conda_env = env_name
    if conda_env is None:
        conda_env = _get_active_conda_env()
    if conda_env is None:
        raise SandboxExecutionError("No conda environment found")

    if shutil.which("conda") is None:
        raise SandboxExecutionError(
            "conda command not found, make sure it is installed on your system and in your PATH"
        )

    with _temp_script_file(script_source) as script_file:
        # conda run removes the need to activate the environment, as would typically be done in an interactive shell
        return subprocess.run(
            [
                "conda",
                "run",
                "-n",
                conda_env,
                "--no-capture-output",
                "python3",
                script_file,
            ],
            capture_output=True,
            text=True,
        )


def _execute_with_system_python(script_source: str) -> subprocess.CompletedProcess:
    python_executable = (
        shutil.which("python3") or shutil.which("python") or sys.executable
    )
    if not python_executable:
        raise SandboxExecutionError("No python executable found")

    with _temp_script_file(script_source) as script_file:
        return subprocess.run(
            [python_executable, script_file], capture_output=True, text=True
        )


class ExecutionEnvironmentType(Enum):
    AUTO_DETECT = 0  # auto-detect the current python interpreter by looking for an active virtual environment
    VENV = 1  # use the python interpreter specified by a venv environment
    CONDA = 2  # use the python interpreter specified by a conda environment
    SYSTEM_DEFAULT = 3  # use the system's default python interpreter


def execute_script_in_sandbox(
    script_source: str,
    env_type: ExecutionEnvironmentType = ExecutionEnvironmentType.AUTO_DETECT,
    **kwargs,
) -> subprocess.CompletedProcess:
    """
    Executes a python script in a sandboxed environment, and returns its output. The script is executed in a different
    process. The execution environment is determined by the `env_type` argument. By default, the logic will attempt
    to auto-detect the correct environment by looking for an active venv or conda environment. If none can be found, it
    will use the system's default python executable, as determined by the user's path. As a last resort, the current
    python execution environment will be used (still in a subprocess).

    Parameters:
        script_source (str): The python script to be executed, as a string.
        env_type: The type of execution environment to use (default: ExecutionEnvironmentType.AUTO_DETECT).
        **kwargs: Additional keyword arguments used by specific execution environments, as follows:
            - venv environments accept a 'path' argument to specify the venv root directory.
            - conda environments accept a 'name' argument to specify the name of the conda environment to use.
    Returns:
        A subprocess.CompletedProcess object containing the output of the script, if any.
    """
    if env_type == ExecutionEnvironmentType.AUTO_DETECT:
        if _is_venv_active():
            return _execute_in_venv(script_source)
        elif _is_conda_active():
            return _execute_in_conda_env(script_source)
        else:
            return _execute_with_system_python(script_source)
    elif env_type == ExecutionEnvironmentType.VENV:
        return _execute_in_venv(script_source, kwargs.get("path"))
    elif env_type == ExecutionEnvironmentType.CONDA:
        return _execute_in_conda_env(script_source, kwargs.get("name"))
    else:
        return _execute_with_system_python(script_source)
