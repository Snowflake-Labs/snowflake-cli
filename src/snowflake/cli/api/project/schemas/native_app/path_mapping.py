from __future__ import annotations

from typing import List, Literal, Optional, Union

from pydantic import Field
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel

# Limited to python options only for now.
# Also, processor name can only be "snowpark" for now.
VirtualEnvTypeOptions = Literal["conda", "venv"]


class VirtualEnvironment(UpdatableModel):
    name: str = Field(
        title="Name of the virtual environment to run the processor in.",
    )
    env_type: VirtualEnvTypeOptions = Field(
        title="Type of the virtual environment to activate.",
    )


class AnnotationProcessor(UpdatableModel):
    name: str = Field(
        title="Name of the processor to be invoked to discover the annotated code."
    )
    virtual_env: Optional[VirtualEnvironment] = Field(
        title="The virtual environment to run the processor in. If none is provided, then the Snowflake CLI will try to get the current environment information.",
        default=None,
    )


class PathMapping(UpdatableModel):
    src: str
    dest: Optional[str] = None
    processors: Optional[List[Union[str, AnnotationProcessor]]] = None
