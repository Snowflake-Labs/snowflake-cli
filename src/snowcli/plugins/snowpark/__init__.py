from snowcli.plugins.snowpark.commands import app
from snowcli.plugins.snowpark.package.commands import app as package_app
from snowcli.plugins.snowpark.procedure_coverage.commands import (
    app as procedure_coverage_app,
)

app.add_typer(package_app)
app.add_typer(procedure_coverage_app)
