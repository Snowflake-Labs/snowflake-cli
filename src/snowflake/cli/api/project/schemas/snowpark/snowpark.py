from __future__ import annotations

from typing import List

from pydantic import Field
from snowflake.cli.api.project.schemas.snowpark.callable import (
    FunctionSchema,
    ProcedureSchema,
)
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel


class Snowpark(UpdatableModel):
    project_name: str = Field(title="Project identifier")
    stage_name: str = Field(title="Stage in which projectâ€™s artifacts will be stored")
    src: str = Field(title="Folder where your code should be located")
    functions: List[FunctionSchema] | None = Field(
        title="List of functions defined in the project", default=[]
    )
    procedures: List[ProcedureSchema] | None = Field(
        title="List of procedures defined in the project", default=[]
    )
