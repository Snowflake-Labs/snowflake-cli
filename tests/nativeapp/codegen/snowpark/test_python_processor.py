from __future__ import annotations

import copy
import subprocess
from pathlib import Path
from textwrap import dedent
from unittest import mock

import pytest
from snowflake.cli.api.project.schemas.native_app.path_mapping import ProcessorMapping
from snowflake.cli.plugins.nativeapp.codegen.sandbox import (
    ExecutionEnvironmentType,
    SandboxExecutionError,
)
from snowflake.cli.plugins.nativeapp.codegen.snowpark.python_processor import (
    SnowparkAnnotationProcessor,
    _determine_virtual_env,
    _execute_in_sandbox,
    edit_setup_script_with_exec_imm_sql,
    generate_create_sql_ddl_statement,
    generate_grant_sql_ddl_statements,
)

from tests.testing_utils.files_and_dirs import temp_local_dir

PROJECT_ROOT = Path("/path/to/project")

# --------------------------------------------------------
# ------------- _determine_virtual_env -------------------
# --------------------------------------------------------


@pytest.mark.parametrize(
    "input_param, expected",
    [
        [ProcessorMapping(name="dummy", properties={"random": "random"}), {}],
        [ProcessorMapping(name="dummy", properties={"env": {"random": "random"}}), {}],
        [
            ProcessorMapping(
                name="dummy",
                properties={"env": {"type": "conda", "name": "snowpark-dev"}},
            ),
            {"env_type": ExecutionEnvironmentType.CONDA, "name": "snowpark-dev"},
        ],
        [
            ProcessorMapping(
                name="dummy", properties={"env": {"type": "venv", "path": "some/path"}}
            ),
            {
                "env_type": ExecutionEnvironmentType.VENV,
                "path": PROJECT_ROOT / "some/path",
            },
        ],
        [
            ProcessorMapping(
                name="dummy", properties={"env": {"type": "venv", "path": "/some/path"}}
            ),
            {"env_type": ExecutionEnvironmentType.VENV, "path": Path("/some/path")},
        ],
        [
            ProcessorMapping(name="dummy", properties={"env": {"type": "current"}}),
            {"env_type": ExecutionEnvironmentType.CURRENT},
        ],
        [ProcessorMapping(name="dummy", properties={"env": {"type": "other"}}), {}],
        [
            ProcessorMapping(name="dummy", properties={"env": {"type": "conda"}}),
            {"env_type": ExecutionEnvironmentType.CONDA, "name": None},
        ],
        [
            ProcessorMapping(name="dummy", properties={"env": {"type": "venv"}}),
            {"env_type": ExecutionEnvironmentType.VENV, "path": None},
        ],
    ],
)
def test_determine_virtual_env(input_param, expected):
    actual = _determine_virtual_env(project_root=PROJECT_ROOT, processor=input_param)
    assert actual == expected


# --------------------------------------------------------
# --------------- _execute_in_sandbox --------------------
# --------------------------------------------------------


@mock.patch(
    "snowflake.cli.plugins.nativeapp.codegen.snowpark.python_processor.execute_script_in_sandbox"
)
def test_execute_in_sandbox_full_entity(mock_sandbox):
    mock_completed_process = mock.MagicMock(spec=subprocess.CompletedProcess)
    mock_completed_process.stdout = '[{"name": "john"}, {"name": "jane"}]'
    mock_completed_process.stderr = ""
    mock_completed_process.returncode = 0
    mock_sandbox.return_value = mock_completed_process

    entity = _execute_in_sandbox(
        py_file="some_file", deploy_root=Path("some/path"), kwargs={}
    )
    assert entity == [{"name": "john"}, {"name": "jane"}]


@mock.patch(
    "snowflake.cli.plugins.nativeapp.codegen.snowpark.python_processor.execute_script_in_sandbox"
)
def test_execute_in_sandbox_all_possible_none_cases(mock_sandbox):
    mock_completed_process = mock.MagicMock(spec=subprocess.CompletedProcess)
    mock_completed_process.returncode = 1
    mock_completed_process.stderr = "DUMMY"
    mock_completed_process.stdout = "DUMMY"
    mock_sandbox.return_value = mock_completed_process
    assert (
        _execute_in_sandbox(
            py_file="some_file", deploy_root=Path("some/path"), kwargs={}
        )
        is None
    )

    mock_completed_process.returncode = 0
    mock_completed_process.stderr = "DUMMY"
    mock_completed_process.stdout = '[{"name": "john"}, {"name": "jane"}'  # Bad input
    assert (
        _execute_in_sandbox(
            py_file="some_file", deploy_root=Path("some/path"), kwargs={}
        )
        is None
    )

    mock_sandbox.side_effect = SandboxExecutionError("dummy error")
    assert (
        _execute_in_sandbox(
            py_file="some_file", deploy_root=Path("some/path"), kwargs={}
        )
        is None
    )

    mock_sandbox.side_effect = ValueError("dummy error")
    assert (
        _execute_in_sandbox(
            py_file="some_file", deploy_root=Path("some/path"), kwargs={}
        )
        is None
    )


# --------------------------------------------------------
# ------- generate_create_sql_ddl_statement -------------
# --------------------------------------------------------


def test_generate_create_sql_ddl_statements_w_all_entries(
    native_app_extension_function, snapshot
):
    assert generate_create_sql_ddl_statement(native_app_extension_function) == snapshot


def test_generate_create_sql_ddl_statements_w_select_entries(
    native_app_extension_function, snapshot
):
    native_app_extension_function.imports = None
    native_app_extension_function.packages = None
    native_app_extension_function.schema_name = None
    native_app_extension_function.secrets = None
    native_app_extension_function.external_access_integrations = None
    assert generate_create_sql_ddl_statement(native_app_extension_function) == snapshot


def test_generate_create_sql_ddl_statements_w_existing_snowpark_dependency(
    native_app_extension_function, snapshot
):
    native_app_extension_function.packages = ["snowflake-snowpark-python"]
    assert generate_create_sql_ddl_statement(native_app_extension_function) == snapshot

    native_app_extension_function.packages = ["snowflake-snowpark-python==0.15.0"]
    assert generate_create_sql_ddl_statement(native_app_extension_function) == snapshot


# --------------------------------------------------------
# ------- generate_grant_sql_ddl_statements --------------
# --------------------------------------------------------


def test_generate_grant_sql_ddl_statements(native_app_extension_function, snapshot):
    assert generate_grant_sql_ddl_statements(native_app_extension_function) == snapshot


# --------------------------------------------------------
# --------- edit_setup_script_with_exec_imm_sql ----------
# --------------------------------------------------------


def test_edit_setup_script_with_exec_imm_sql(snapshot):
    manifest_contents = dedent(
        f"""\
        manifest_version: 1

        artifacts:
            setup_script: moduleA/moduleC/setup.sql
        """
    )
    dir_structure = {
        "output/deploy/manifest.yml": manifest_contents,
        "output/deploy/moduleA/moduleC/setup.sql": "create application role app_public;",
        "output/deploy/__generated/moduleB/dummy.sql": "#this is a file",
        "output/deploy/__generated/dummy.sql": "#this is a file",
    }

    with temp_local_dir(dir_structure=dir_structure) as local_path:
        deploy_root = Path(local_path, "output", "deploy")
        generated_root = Path(deploy_root, "__generated")
        collected_sql_files = [
            Path(generated_root, "moduleB", "dummy.sql"),
            Path(generated_root, "dummy.sql"),
        ]

        edit_setup_script_with_exec_imm_sql(
            collected_sql_files=collected_sql_files,
            deploy_root=deploy_root,
            generated_root=generated_root,
        )

        main_file = Path(deploy_root, "__generated", "__generated.sql")
        assert main_file.is_file()
        with open(main_file, "r") as f:
            assert f.read() == snapshot

        setup_file = Path(deploy_root, "moduleA", "moduleC", "setup.sql")
        with open(setup_file, "r") as f:
            assert f.read() == snapshot


def test_edit_setup_script_with_exec_imm_sql_noop(snapshot):
    manifest_contents = dedent(
        f"""\
        manifest_version: 1

        artifacts:
            setup_script: moduleA/moduleC/setup.sql
        """
    )
    dir_structure = {
        "output/deploy/manifest.yml": manifest_contents,
        "output/deploy/moduleA/moduleC/setup.sql": None,
        "output/deploy/__generated/__generated.sql": "#some text",
    }

    with temp_local_dir(dir_structure=dir_structure) as local_path:
        deploy_root = Path(local_path, "output", "deploy")
        collected_sql_files = [
            Path(deploy_root, "__generated", "dummy.sql"),
        ]
        edit_setup_script_with_exec_imm_sql(
            collected_sql_files=collected_sql_files,
            deploy_root=deploy_root,
            generated_root=Path(deploy_root, "__generated"),
        )
        main_file = Path(deploy_root, "__generated", "__generated.sql")
        assert main_file.is_file()
        with open(main_file, "r") as f:
            assert f.read() == snapshot

        dummy_file = Path(deploy_root, "__generated", "dummy.sql")
        assert not dummy_file.exists()


def test_edit_setup_script_with_exec_imm_sql_symlink(snapshot):
    manifest_contents = dedent(
        f"""\
        manifest_version: 1

        artifacts:
            setup_script: setup.sql
        """
    )
    dir_structure = {
        "setup.sql": "create application role admin;",
        "output/deploy/manifest.yml": manifest_contents,
    }

    with temp_local_dir(dir_structure=dir_structure) as local_path:
        deploy_root = Path(local_path, "output", "deploy")

        deploy_root_setup_script = Path(deploy_root, "setup.sql")
        deploy_root_setup_script.symlink_to(Path(local_path, "setup.sql"))

        generated_root = Path(deploy_root, "__generated")
        collected_sql_files = [
            Path(generated_root, "moduleB", "dummy.sql"),
            Path(generated_root, "dummy.sql"),
        ]
        edit_setup_script_with_exec_imm_sql(
            collected_sql_files=collected_sql_files,
            deploy_root=deploy_root,
            generated_root=Path(deploy_root, "__generated"),
        )

        main_file = Path(deploy_root, "__generated", "__generated.sql")
        assert main_file.is_file()
        with open(main_file, "r") as f:
            assert f.read() == snapshot

        # Should not be a symlink anymore
        assert not deploy_root_setup_script.is_symlink()

        with open(deploy_root_setup_script, "r") as f:
            assert f.read() == snapshot

        with open(Path(local_path, "setup.sql"), "r") as f:
            assert f.read() == snapshot


# --------------------------------------------------------
# ------------- SnowparkAnnotationProcessor --------------
# --------------------------------------------------------
manifest_contents = dedent(
    f"""\
    manifest_version: 1

    artifacts:
        setup_script: moduleA/moduleC/setup.sql
    """
)

minimal_dir_structure = {
    "a/b/c/main.py": "# this is a file\n",
    "a/b/c/data.py": "# this is a file\n",
    "output/deploy": None,
    "output/deploy/manifest.yml": manifest_contents,
    "output/deploy/moduleA/moduleC/setup.sql": "create application role app_public;",
    "output/deploy/stagepath/main.py": "# this is a file\n",
    "output/deploy/stagepath/data.py": "# this is a file\n",
    "output/deploy/stagepath/extra_import1.zip": None,
    "output/deploy/stagepath/extra_import2.zip": None,
}


@mock.patch(
    "snowflake.cli.plugins.nativeapp.codegen.snowpark.python_processor._execute_in_sandbox",
)
def test_process_no_collected_functions(
    mock_sandbox, native_app_project_instance, snapshot
):
    with temp_local_dir(minimal_dir_structure) as local_path:
        native_app_project_instance.native_app.artifacts = [
            {"src": "a/b/c/*.py", "dest": "stagepath/", "processors": ["SNOWPARK"]}
        ]
        mock_sandbox.side_effect = [None, []]
        deploy_root = Path(local_path, "output/deploy")
        generated_root = Path(deploy_root, "__generated")
        output = SnowparkAnnotationProcessor(
            project_definition=native_app_project_instance,
            project_root=local_path,
            deploy_root=deploy_root,
            generated_root=generated_root,
        ).process(
            artifact_to_process=native_app_project_instance.native_app.artifacts[0],
            processor_mapping=ProcessorMapping(name="SNOWPARK"),
            write_to_sql=False,  # For testing
        )
        assert output == snapshot
        assert not Path(generated_root, "stagepath/main.sql").exists()
        assert not Path(generated_root, "stagepath/data.sql").exists()


@mock.patch(
    "snowflake.cli.plugins.nativeapp.codegen.snowpark.python_processor._execute_in_sandbox",
)
def test_process_with_collected_functions(
    mock_sandbox,
    native_app_project_instance,
    native_app_extension_function_raw_data,
    snapshot,
):

    with temp_local_dir(minimal_dir_structure) as local_path:
        imports_variation = copy.deepcopy(native_app_extension_function_raw_data)
        imports_variation["imports"] = [
            "@dummy_stage_str",
            "/",
            "stagepath/extra_import1.zip",
            "stagepath/some_dir_str",
            "/stagepath/withslash.py",
            "stagepath/data.py",
        ]
        processor_mapping = ProcessorMapping(
            name="snowpark",
            properties={"env": {"type": "conda", "name": "snowpark-dev"}},
        )
        native_app_project_instance.native_app.artifacts = [
            {
                "src": "a/b/c/*.py",
                "dest": "stagepath/",
                "processors": [processor_mapping],
            }
        ]
        mock_sandbox.side_effect = [
            [native_app_extension_function_raw_data],
            [imports_variation],
        ]
        deploy_root = Path(local_path, "output/deploy")
        generated_root = Path(deploy_root, "__generated")
        output = SnowparkAnnotationProcessor(
            project_definition=native_app_project_instance,
            project_root=local_path,
            deploy_root=deploy_root,
            generated_root=generated_root,
        ).process(
            artifact_to_process=native_app_project_instance.native_app.artifacts[0],
            processor_mapping=processor_mapping,
        )
        assert output == snapshot

        main_file = Path(deploy_root, "__generated", "__generated.sql")
        assert main_file.is_file()
        with open(main_file, "r") as f:
            assert f.read() == snapshot

        # Should have execute imm
        with open(
            Path(local_path, "output/deploy/moduleA/moduleC/setup.sql"), "r"
        ) as f:
            assert f.read() == snapshot

        with open(
            Path(local_path, "output/deploy/__generated/stagepath/main.sql"), "r"
        ) as f:
            assert f.read() == snapshot

        with open(
            Path(local_path, "output/deploy/__generated/stagepath/data.sql"), "r"
        ) as f:
            assert f.read() == snapshot
