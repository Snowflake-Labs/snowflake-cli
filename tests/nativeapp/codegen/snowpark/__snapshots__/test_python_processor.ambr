# serializer version: 1
# name: test_create_and_write_to_sql_files_no_existing_generate
  dict({
    PosixPath('moduleA/main.py'): PosixPath('__generated/moduleA/main.sql'),
    PosixPath('moduleB/main.py'): PosixPath('__generated/moduleB/main.sql'),
  })
# ---
# name: test_edit_setup_script_with_exec_imm_sql
  '''
  EXECUTE IMMEDIATE FROM '/__generated/moduleB/dummy.sql';
  EXECUTE IMMEDIATE FROM '/__generated/dummy.sql';
  
  '''
# ---
# name: test_edit_setup_script_with_exec_imm_sql.1
  '''
  create application role app_public;
  EXECUTE IMMEDIATE FROM '/__generated/__generated.sql';
  
  '''
# ---
# name: test_generate_create_sql_ddl_statements_w_all_entries
  '''
  CREATE OR REPLACE
  PROCEDURE DATA.my_function(first int DEFAULT 42)
  RETURNS int
  LANGUAGE PYTHON
  RUNTIME_VERSION=3.11
  IMPORTS=('/path/to/import1.py', '/path/to/import2.zip')
  PACKAGES=('package_one==1.0.2', 'package_two')
  EXTERNAL_ACCESS_INTEGRATIONS=('integration_one', 'integration_two')
  SECRETS=('key1'=secret_one, 'key2'=integration_two)
  HANDLER='my_function_handler'
  EXECUTE AS OWNER;
  
  '''
# ---
# name: test_generate_create_sql_ddl_statements_w_select_entries
  '''
  CREATE OR REPLACE
  PROCEDURE my_function(first int DEFAULT 42)
  RETURNS int
  LANGUAGE PYTHON
  RUNTIME_VERSION=3.11
  HANDLER='my_function_handler'
  EXECUTE AS OWNER;
  
  '''
# ---
# name: test_generate_grant_sql_ddl_statements
  '''
  GRANT USAGE ON PROCEDURE DATA.my_function
  TO APPLICATION ROLE APP_ADMIN;
  GRANT USAGE ON PROCEDURE DATA.my_function
  TO APPLICATION ROLE APP_VIEWER;
  '''
# ---
# name: test_process_no_collected_functions
  ''
# ---
# name: test_process_with_collected_functions
  '''
  -- stagepath/main.py
  CREATE OR REPLACE
  PROCEDURE DATA.my_function(first int DEFAULT 42)
  RETURNS int
  LANGUAGE PYTHON
  RUNTIME_VERSION=3.11
  IMPORTS=('/path/to/import1.py', '/path/to/import2.zip', '/stagepath/main.py')
  PACKAGES=('package_one==1.0.2', 'package_two')
  EXTERNAL_ACCESS_INTEGRATIONS=('integration_one', 'integration_two')
  SECRETS=('key1'=secret_one, 'key2'=integration_two)
  HANDLER='main.my_function_handler'
  EXECUTE AS OWNER;
  
  GRANT USAGE ON PROCEDURE DATA.my_function
  TO APPLICATION ROLE APP_ADMIN;
  GRANT USAGE ON PROCEDURE DATA.my_function
  TO APPLICATION ROLE APP_VIEWER;
  -- stagepath/data.py
  CREATE OR REPLACE
  PROCEDURE DATA.my_function(first int DEFAULT 42)
  RETURNS int
  LANGUAGE PYTHON
  RUNTIME_VERSION=3.11
  IMPORTS=('/path/to/import1.py', '/path/to/import2.zip', '/stagepath/data.py')
  PACKAGES=('package_one==1.0.2', 'package_two')
  EXTERNAL_ACCESS_INTEGRATIONS=('integration_one', 'integration_two')
  SECRETS=('key1'=secret_one, 'key2'=integration_two)
  HANDLER='data.my_function_handler'
  EXECUTE AS OWNER;
  
  GRANT USAGE ON PROCEDURE DATA.my_function
  TO APPLICATION ROLE APP_ADMIN;
  GRANT USAGE ON PROCEDURE DATA.my_function
  TO APPLICATION ROLE APP_VIEWER;
  '''
# ---
